### ISSUE TITLE
Which cleaner to use for French ?

### ISSUE THREAD

**Issue by Ca-ressemble-a-du-fake** [CorentinJ/Real-Time-Voice-Cloning#894]
Hi,

I noticed that there are 3 cleaners for tacotron. Before I noticed that, I tried the `english_cleaners` and it went fast. Then I tried the `transliteration_cleaners` but it went half the speed of the english one. Now I see there is also a `basic_cleaners`. I understand that if I've updated the list of symbols I should use this `basic_cleaners` to keep input text as it is (with accents and co). But I am not sure which one I should use. 

Moreover if I change the cleaner do I have to run the `preprocess_audio` and `preprocess_embeds` scripts again, or the training can start with the existing preprocessed datasets ?

**Comment by blue-fish** [comment]
You should start with `basic_cleaners` and add functionality to handle numerals and abbreviations.

There is no need to rerun preprocess after updating cleaners. The cleaners intercept and replace text in real time as it is input to the model.

**Comment by blue-fish** [comment]
Here is how a different TTS does it for French. They train on phonemes instead of graphemes like this repo does.

https://github.com/coqui-ai/TTS/blob/33aa27e2d6a80d38a7ea8928ad14d032d2c53a40/TTS/tts/utils/text/cleaners.py#L103-L110

https://github.com/coqui-ai/TTS/blob/33aa27e2d6a80d38a7ea8928ad14d032d2c53a40/TTS/tts/utils/text/abbreviations.py#L29

**Comment by Ca-ressemble-a-du-fake** [comment]
Thank you very much! I will give the basic cleaners a try. Your answer is very valuable, I understand better how the text processing works.

**Comment by Ca-ressemble-a-du-fake** [comment]
By the way is there a way to test and insure the cleaner works as expected without training ?

**Comment by Ca-ressemble-a-du-fake** [comment]
I tried `basic_cleaners` but I read weird inputs in the console :

> 
+----------------+------------+---------------+------------------+
| Steps with r=7 | Batch Size | Learning Rate | Outputs/Step (r) |
+----------------+------------+---------------+------------------+
|   20k Steps    |     20     |     0.001     |        7         |
+----------------+------------+---------------+------------------+
 
{| Epoch: 2/44 (39/461) | Loss: 1.054 | 0.99 steps/s | Step: 0k | }**Input at step 500:     ,     é,   é  -ê     û   .~___________**
{| Epoch: 3/44 (78/461) | Loss: 0.9771 | 1.0 steps/s | Step: 1k | }**Input at step 1000: -   éé,  éà .~___________**
{| Epoch: 4/44 (117/461) | Loss: 0.9348 | 1.0 steps/s | Step: 1k | }**Input at step 1500:    é     éê.~__________________________**
{| Epoch: 5/44 (156/461) | Loss: 0.8881 | 1.0 steps/s | Step: 2k | }**Input at step 2000:          ,       . ~_____________________**
{| Epoch: 6/44 (195/461) | Loss: 0.8642 | 1.0 steps/s | Step: 2k | }**Input at step 2500:  -à ,  é       , î    ,      é.~**
{| Epoch: 7/44 (234/461) | Loss: 0.8479 | 1.0 steps/s | Step: 3k | }**Input at step 3000:   ,      è     .~__________________________**
...
{| Epoch: 10/44 (351/461) | Loss: 0.8224 | 1.0 steps/s | Step: 4k | }**Input at step 4500:  é é      é    ê    â**    

Previously, with the english_cleaners the inputs were legible (only accented letters were removed). When I test the cleaners with an input, they return the expected output eg :
 
`print(basic_cleaners("J’attendais, suivant la coutume, que la phrase quotidienne fût prononcée."))`
> j’attendais, suivant la coutume, que la phrase quotidienne fût prononcée.

`print(english_cleaners("J’attendais, suivant la coutume, que la phrase quotidienne fût prononcée."))`
> j'attendais, suivant la coutume, que la phrase quotidienne fut prononcee.

THere is no `~` in train.txt so where do all those ~ come from ?

What is actually going on, will it spoil the training ?

**Comment by blue-fish** [comment]
> THere is no `~` in train.txt so where do all those ~ come from ?

It is the end of sequence marker. It helps train the stop prediction.
https://github.com/CorentinJ/Real-Time-Voice-Cloning/blob/7432046efc23cabf176f9fdc8d2fd67020059478/synthesizer/utils/text.py#L38-L39

**Comment by Ca-ressemble-a-du-fake** [comment]
My bad, I forgot all lower case consonants in symbols.py.

**Comment by Ca-ressemble-a-du-fake** [comment]
Thanks @blue-fish . And is it expected to have more "_" (called `_pad` in `symbols.py`) than characters in the input ? Eg :

> Input at step 14000: nul ne doute quil faille muscler la réforme bancaire.~___________________________________________________________________________________________________________________________________________________

The input is 53 chars long whereas there are 147 "_" after ~

One more thing. As you can see the ' (apostrophe) is filtered out (expected should be "qu**'**il" instead of "quil"). I am not sure it is an issue because apostrophe has no particular sound but it should be present since it is in `symbols.py` :

`_characters = "ABCDEFGHIJKLMNOPQRSTUVWXYZŸÜÛÙŒÔÏÎËÊÈÉÇÆÂÀabcdefghijklmnopqrstuvwxyzàâæçéèêëîïôœùûüÿ!\'\"(),-.:;? "`

If I just run a test against this pipeline the apostrophe is kept, so it looks like the apostrophe is removed somewhere else than in the cleaners. Can you tell me if my cleaner works correctly regarding the apostrophe ?

**Comment by blue-fish** [comment]
> Thanks @blue-fish . And is it expected to have more "_" (called `_pad` in `symbols.py`) than characters in the input ? Eg :
> 
> > Input at step 14000: nul ne doute quil faille muscler la réforme bancaire.~___________________________________________________________________________________________________________________________________________________
> 
> The input is 53 chars long whereas there are 147 "_" after ~

Text padding with `_` is used so all inputs in the batch have the same length. This allows the texts to be batched together and processed at the same time.


> One more thing. As you can see the ' (apostrophe) is filtered out (expected should be "qu**'**il" instead of "quil"). I am not sure it is an issue because apostrophe has no particular sound but it should be present since it is in `symbols.py` :
> 
> `_characters = "ABCDEFGHIJKLMNOPQRSTUVWXYZŸÜÛÙŒÔÏÎËÊÈÉÇÆÂÀabcdefghijklmnopqrstuvwxyzàâæçéèêëîïôœùûüÿ!\'\"(),-.:;? "`
> 
> If I just run a test against this pipeline the apostrophe is kept, so it looks like the apostrophe is removed somewhere else than in the cleaners. Can you tell me if my cleaner works correctly regarding the apostrophe ?

What is most likely going on here, is your apostrophe in `train.txt` is using the `’` character. Notice how it is different from the apostrophe in symbols.py. If you use basic_cleaners, it will be missing the `convert_to_ascii` step which was converting `’` to `'`. When [symbols_to_sequence](https://github.com/CorentinJ/Real-Time-Voice-Cloning/blob/7432046efc23cabf176f9fdc8d2fd67020059478/synthesizer/utils/text.py#L32) encounters a character that is not in symbols, it is dropped.

If there is a problem that prevents you from using convert_to_ascii, then you should either find-and-replace `’` to `'` in train.txt, or update your symbols.py to include `’`.

**Comment by Ca-ressemble-a-du-fake** [comment]
Thank you @blue-fish you spotted it! Indeed in Siwis dataset, they use both apostroph signs (actually  `'` around 100 times, and  `’` 6k+ times). `convert_to_ascii` does not keep accented letters so I'll add it to `symbols.py` and use the `basic_cleaner` as recommended [by tacotron team](https://github.com/keithito/tacotron/blob/master/TRAINING_DATA.md).

By the way what's the name of `’` and how do you type it (I had to paste it from you message inside `symbols.py`!) ?

**Comment by blue-fish** [comment]
The symbol `’` has many names. The unicode name is "right single quotation mark", but it is sometimes called a curved quote, curly quote, or smart quote.

**Comment by Ca-ressemble-a-du-fake** [comment]
Thank you @blue-fish, I did not find it on my keyboard and firefox treats both signs as the same apostoph !

**Comment by Ca-ressemble-a-du-fake** [comment]
Thank you @blue-fish, I did not find it on my keyboard and firefox treats both signs as the
